Name: Aryan Jha

Course: MBA TECH AI

Topic: Feed Forwarding

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [ ]:
#Input layer
df=pd.read_csv('/content/Iris (1).csv')

In [ ]:
species=pd.get_dummies(df['Species'])

In [ ]:
species.head()

Iris-setosa  Iris-versicolor  Iris-virginica
0            1                0               0
1            1                0               0
2            1                0               0
3            1                0               0
4            1                0               0

In [ ]:
df.drop('Species',axis=1,inplace=True)

In [ ]:
df

SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0              5.1           3.5            1.4           0.2
1              4.9           3.0            1.4           0.2
2              4.7           3.2            1.3           0.2
3              4.6           3.1            1.5           0.2
4              5.0           3.6            1.4           0.2
..             ...           ...            ...           ...
145            6.7           3.0            5.2           2.3
146            6.3           2.5            5.0           1.9
147            6.5           3.0            5.2           2.0
148            6.2           3.4            5.4           2.3
149            5.9           3.0            5.1           1.8

[150 rows x 4 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler
std_scaler = StandardScaler()

In [ ]:
df_scaled=std_scaler.fit_transform(df)

In [ ]:
df_transformed=pd.DataFrame(df_scaled,columns=['SepalLengthCm','SepalWidthCm','PetalLengthCm','PetalWidthCm'])

In [ ]:
Iris_df=pd.concat([df_transformed,species], axis=1)
#Axis=1 to concatenate columns

Input dimensions: 150 (no. of samples) x 4 (no. of features)

In [ ]:
input=Iris_df.iloc[:, 0:4]

In [ ]:
input.shape

(150, 4)

In [ ]:
print(input)

     SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0        -0.900681      1.032057      -1.341272     -1.312977
1        -1.143017     -0.124958      -1.341272     -1.312977
2        -1.385353      0.337848      -1.398138     -1.312977
3        -1.506521      0.106445      -1.284407     -1.312977
4        -1.021849      1.263460      -1.341272     -1.312977
..             ...           ...            ...           ...
145       1.038005     -0.124958       0.819624      1.447956
146       0.553333     -1.281972       0.705893      0.922064
147       0.795669     -0.124958       0.819624      1.053537
148       0.432165      0.800654       0.933356      1.447956
149       0.068662     -0.124958       0.762759      0.790591

[150 rows x 4 columns]


Initialize weights and biases

In [ ]:
np.random.seed(42) #To ensure we get the same weights every time we run it
w_i_h1=np.random.rand(4,1)
w_i_h2=np.random.rand(4,1)
w_i_h3=np.random.rand(4,1)
w_h_o1=np.random.rand(2,1)
w_h_o2=np.random.rand(2,1)
w_h_o3=np.random.rand(2,1)
bias1=np.random.rand(1)
bias2=np.random.rand(1)
bias3=np.random.rand(1)

In [ ]:
print(w_i_h1)
print(w_i_h2)
print(w_i_h3)
print(w_h_o1)

[[0.37454012]
 [0.95071431]
 [0.73199394]
 [0.59865848]]
[[0.15601864]
 [0.15599452]
 [0.05808361]
 [0.86617615]]
[[0.60111501]
 [0.70807258]
 [0.02058449]
 [0.96990985]]
[[0.83244264]
 [0.21233911]]


Sigmoid function

In [ ]:
import math

In [ ]:
sigmoid= lambda x: 1/(1+(np.exp(-x)))

In [ ]:
def sigmoid(x):
  return 1/(1+(np.exp(-x)))

In [ ]:
sigmoid(0)

0.5

In [ ]:
print(np.dot(input,w_i_h1))

[[-1.12397760e+00]
 [-2.31473260e+00]
 [-2.00712619e+00]
 [-2.18925583e+00]
 [-9.49361715e-01]
 [ 1.74452398e-01]
 [-1.49217944e+00]
 [-1.34773258e+00]
 [-2.76164185e+00]
 [-2.13181655e+00]
 [-5.06209341e-01]
 [-1.39687169e+00]
 [-2.43882224e+00]
 [-2.79060943e+00]
 [ 2.10437850e-01]
 [ 1.32733893e+00]
 [ 7.95099471e-03]
 [-1.04527019e+00]
 [ 1.18935653e-02]
 [-3.43650508e-01]
 [-1.08295297e+00]
 [-4.84941208e-01]
 [-1.29739203e+00]
 [-1.20297554e+00]
 [-1.27199564e+00]
 [-2.18609967e+00]
 [-1.14869241e+00]
 [-1.03697002e+00]
 [-1.29859348e+00]
 [-1.88225014e+00]
 [-2.05686602e+00]
 [-1.00878885e+00]
 [ 2.04311231e-01]
 [ 5.97538085e-01]
 [-2.13181655e+00]
 [-1.91260485e+00]
 [-9.84074033e-01]
 [-2.13181655e+00]
 [-2.58326909e+00]
 [-1.30235035e+00]
 [-1.13227777e+00]
 [-3.99916622e+00]
 [-2.14327288e+00]
 [-7.71279487e-01]
 [-9.84416957e-02]
 [-2.28140742e+00]
 [-3.80732566e-01]
 [-2.01088307e+00]
 [-5.51591569e-01]
 [-1.60935604e+00]
 [ 1.39641590e+00]
 [ 1.11957924e+00]
 [ 1.2929936

In [ ]:
print(bias1)

[0.43194502]


In [ ]:
#Feedforward Step1 - input to hidden layer
Z2_1=np.dot(input,w_i_h1) + bias1
Z2_2=np.dot(input,w_i_h2) + bias2
Z2_3=np.dot(input,w_i_h3) + bias3

In [ ]:
#Feedforward step - 2
h2_1=sigmoid(Z2_1)
h2_2= sigmoid(Z2_2)
h2_3= sigmoid(Z2_3)

In [ ]:
h2_1.shape

(150, 1)

If we don't apply a sigmoid function, it will act only as a linear classifier.
The activation function is added to include non-linearity.

In [ ]:
#Each output neuron gets input from both hidden layers so append both into an array
h2=np.append(h2_1,h2_2,axis=1)

In [ ]:
h2=np.append(h2_2,h2_3,axis=1)

In [ ]:
h2.shape

(150, 2)

In [ ]:
#Feedforward step 3 - input from hidden layer to output
Z3_1=np.dot(h2,w_h_o1)
Z3_2=np.dot(h2,w_h_o2)
Z3_3=np.dot(h2,w_h_o3)

In [ ]:
#Feedforward step 4 - generating outputs
o1=sigmoid(Z3_1)
o2= sigmoid(Z3_2)
o3= sigmoid(Z3_3)

In [ ]:
print(o1.shape)

(150, 1)


In [ ]:
print(Z2_1.shape)

(150, 1)


In [ ]:
t1 = np.array(Iris_df.iloc[:,4])

In [ ]:
t1.shape

(150,)

In [ ]:
t1 = t1.reshape((-1,1))

In [ ]:
t1.shape

(150, 1)

In [ ]:
e1 = t1-o1

In [ ]:
print(e1.shape)

(150, 1)


In [ ]:
t2 = np.array(Iris_df.iloc[:,5])
t2 = t2.reshape((-1,1))
e2 = t2-o2
print(e2.shape)

(150, 1)


In [ ]:
t3 = np.array(Iris_df.iloc[:,6])
t3 = t3.reshape((-1,1))
e3 = t3-o3
print(e3.shape)

(150, 1)


In [ ]:
error = np.multiply(e1,e1)+np.multiply(e2,e2)+np.multiply(e3,e3)

In [ ]:
print(error.shape)

(150, 1)


In [ ]:
error_total = np.sum(error)

In [ ]:
error_total

135.88261448713052